### ISSUE TITLE
USB keyboard not detected on ASUS RS904A

### ISSUE THREAD

**Issue by debrouxl** [memtest86plus/memtest86plus#6]
As described in https://github.com/martinwhitaker/pcmemtest/issues/16 , while the current OHCI / EHCI USB support works on Martin's computers, it doesn't yet work properly on two similar 4S Opteron 62xx & 63xx computers with AMD chipsets I have access to.

The USB detection screen shows that the USB controllers themselves are detected, but the HID keyboard enumeration does not work on one of the two computers, and the USB detection screen can't be exited out of the box. Both of the computers' legacy BIOS support USB legacy keyboard emulation, so passing `keyboard=legacy` through the command line is enough for interactive usage in legacy BIOS mode; one of the two computers also supports UEFI mode and has no PS/2 port, and there's no way to use the memory tester interactively in UEFI mode on that computer, hence the small command line configuration capability improvement in #4 .

```
# lspci | grep USB
00:12.0 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0/SB8x0/SB9x0 USB OHCI0 Controller
00:12.1 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0 USB OHCI1 Controller
00:12.2 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0/SB8x0/SB9x0 USB EHCI Controller
00:13.0 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0/SB8x0/SB9x0 USB OHCI0 Controller
00:13.1 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0 USB OHCI1 Controller
00:13.2 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0/SB8x0/SB9x0 USB EHCI Controller
00:14.5 USB controller: Advanced Micro Devices, Inc. [AMD/ATI] SB7x0/SB8x0/SB9x0 USB OHCI2 Controller
# lspci -n | grep 00:1[234]
00:12.0 0c03: 1002:4397
00:12.1 0c03: 1002:4398
00:12.2 0c03: 1002:4396
00:13.0 0c03: 1002:4397
00:13.1 0c03: 1002:4398
00:13.2 0c03: 1002:4396
00:14.5 0c03: 1002:4399
```

**Comment by martinwhitaker** [comment]
@x86fr I think it's better to start a new issue for each case.

**Comment by x86fr** [comment]
Ok, closing this one. Will open distinct thread for specific issues related to USB handling.

**Comment by debrouxl** [comment]
Alright, I'll test this tomorrow evening or later.

**Comment by debrouxl** [comment]
This "Port %i handover failed" trace added on top of current HEAD a4f1ba37b3b2ad1f5c11e1ac8d2484035bc746ce does not fire on the RS904A. I made the test twice, and checked that I'm using the right binary by changing one of the main display strings, as usual.

**Comment by martinwhitaker** [comment]
And the OHCI controller still didn't detect any devices?

**Comment by debrouxl** [comment]
Nope, neither on the RS904A in BIOS mode, nor on the R815 (on which I did not test commenting out the handover call, oops) in UEFI mode. The R815 still reboots pretty quickly after the main display appears, like it did with pcmemtest 1.6-pre, even with the `nosm` command line parameter.

**Comment by martinwhitaker** [comment]
Which is a regression then, as the screenshot in https://github.com/memtest86plus/memtest86plus/issues/6#issuecomment-1047683939 shows the R815 detecting the keyboard after handover to the OHCI.

**Comment by martinwhitaker** [comment]
I can't see anything else wrong by code inspection, and without a way to reproduce the fault, there's not much more I can do. I've added a `noehci` boot command line option to allow the EHCI to be skipped, which at least provides a workaround.

**Comment by martinwhitaker** [comment]
Please test if this patch makes any difference:
````
--- a/system/ohci.c
+++ b/system/ohci.c
@@ -448,19 +448,9 @@ bool ohci_init(uintptr_t base_addr, usb_hcd_t *hcd)
         fm_interval = 0x2edf;
     }
 
-    // Prepare for host controller setup (see section 5.1.1.3 of the OHCI spec.).
-    switch (read32(&op_regs->control) & OHCI_CTRL_HCFS) {
-      case OHCI_CTRL_HCFS_RST:
-        usleep(50*MILLISEC);
-        break;
-      case OHCI_CTRL_HCFS_SUS:
-      case OHCI_CTRL_HCFS_RES:
-        flush32(&op_regs->control, OHCI_CTRL_HCFS_SUS);
-        usleep(10*MILLISEC);
-        break;
-      default: // operational
-        break;
-    }
+    // Reset the USB bus.
+    flush32(&op_regs->control, OHCI_CTRL_HCFS_RST);
+    usleep(50*MILLISEC);
 
     // Reset the host controller.
     write32(&op_regs->command_status, OHCI_CMD_HCR);
````

**Comment by debrouxl** [comment]
* out of the box, a build of 66c6dfa3576 + this sole patch does not detect the Logitech K120, whether it's plugged to one of the two front ports or one of the two back ports;
* for sanity checking: passing `noehci` on the command line makes the keyboard work well.

**Comment by debrouxl** [comment]
Without much hope, I tested after the patch which fixes #56 was applied: no change.

**Comment by martinwhitaker** [comment]
Reading back through the comments, on the 22nd February your keyboard was detected by the OHCI controller in the R815 after handover (although key presses weren't detected), but on the 30th March it wasn't. There was only one significamt change to ohci.c between those dates (commit 6c79f4d2)  and two to ehci.c (commit b609c4dd and commit e0d42db0). It might help if we knew which of those commits made the difference.

**Comment by martinwhitaker** [comment]
Fixed by commit 3a10701.
